Engine/Results/CapacityEstimate.cs

```diff
@@ -0,0 +1,167 @@
+/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Interfaces;
+using QuantConnect.Orders;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Lean.Engine.Results
+{
+    /// <summary>
+    /// Estimates dollar volume capacity of algorithm using all Symbols in the portfolio.
+    /// </summary>
+    public class CapacityEstimate
+    {
+        private readonly IAlgorithm _algorithm;
+        private readonly Dictionary<Symbol, SymbolCapacity> _capacityBySymbol;
+        private List<SymbolCapacity> _monitoredSymbolCapacity;
+        private HashSet<SymbolCapacity> _monitoredSymbolCapacitySet;
```

Comment:
Why do we need so many collections for the `SymbolCapacity`? Note that the `HashSet<SymbolCapacity>` is using the reference for uniqueness

Reply:
Maybe monitored could directly be a state in the `SymbolCapacity` and avoid having another collection?

Reply:
We have multiple collections to avoid having to perform an O(n) lookup whenever we're wanting to check whether a particular SymbolData instance is being "monitored" (i.e. it is currently aggregating market dollar volume for its capacity calculation). This way we ensure that we're performing a O(1) lookup, which would prevent any sort of looping.

Reply:
Just a note for posterity; it was done to avoid a look-up of a public property on the SymbolCapacity objects on each loop. So we just check if there are any monitored assets; update those, and continue if nothing to do (99%) of the time.